src/parser/shaman/restoration/modules/talents/UnleashLife.js

```diff
@@ -55,128 +66,104 @@ class UnleashLife extends Analyzer {
-      this.healing += calculateEffectiveHealing(event, UNLEASH_LIFE_HEALING_INCREASE);
-    } else if (spellId === SPELLS.UNLEASH_LIFE_TALENT.id) {
-      this.healing += event.amount + (event.absorbed || 0);
+
+      // Chain heal has up to 4 events, setting the variable to -1 to indicate that there might be more events coming
+    } else if (spellId === SPELLS.CHAIN_HEAL.id && (this.unleashLifeHealRemaining > 0 || (this.unleashLifeHealRemaining < 0 && this.buffedChainHealTimestamp + BUFFER_MS > event.timestamp))) {
+      this.healingBuff[spellId].healing += calculateEffectiveHealing(event, UNLEASH_LIFE_HEALING_INCREASE);
+      this.unleashLifeHealRemaining = -1;
+      this.buffedChainHealTimestamp = event.timestamp;
+      debug && console.log("Heal:", spellId);
     }
   }
 
+
   on_byPlayer_cast(event) {
     const spellId = event.ability.guid;
 
-    if(spellId === SPELLS.UNLEASH_LIFE_TALENT.id) {
+    if (spellId === SPELLS.UNLEASH_LIFE_TALENT.id) {
       this.unleashLifeCasts += 1;
-      this.unleashLifeRemaining += 1;
+      this.unleashLifeRemaining = true;
       this.lastUnleashLifeTimestamp = event.timestamp;
+      debug && console.log("New Unleash", event.timestamp);
     }
 
-    const hasUnleashLife = this.selectedCombatant.hasBuff(SPELLS.UNLEASH_LIFE_TALENT.id, event.timestamp, BUFFER_MS, BUFFER_MS);
+    if (this.unleashLifeRemaining && (this.lastUnleashLifeTimestamp + UNLEASH_LIFE_DURATION) <= event.timestamp) {
+      this.unleashLifeRemaining = false;
+      debug && console.log("Cast Timed out", event.timestamp);
+      return;
+    }
 
-    if(!hasUnleashLife) {
+    const hasUnleashLife = this.selectedCombatant.hasBuff(SPELLS.UNLEASH_LIFE_TALENT.id, event.timestamp, BUFFER_MS, BUFFER_MS);
+    if (!hasUnleashLife) {
       return;
     }
 
     if (this.unleashLifeRemaining) {
-      if((this.lastUnleashLifeTimestamp + unleashLifeBuffDuration) <= event.timestamp) {
-        this.unleashLifeRemaining = 0;
-        return;
-      }
-
-      switch(spellId) {
-        case SPELLS.CHAIN_HEAL.id:
-            this.buffedChainHeals += 1;
-            this.unleashLifeRemaining -= 1;
-          break;
-        case SPELLS.HEALING_WAVE.id:
-            this.buffedHealingWaves += 1;
-            this.unleashLifeRemaining -= 1;
-          break;
-        case SPELLS.HEALING_SURGE_RESTORATION.id:
-            this.buffedHealingSurges += 1;
-            this.unleashLifeRemaining -= 1;
-          break;
-        case SPELLS.RIPTIDE.id:
-            this.buffedRiptides += 1;
-            this.unleashLifeRemaining -= 1;
-          break;
-        default:
-          return;
+      if (this.healingBuff[spellId]) {
+        this.healingBuff[spellId].castAmount += 1;
+        this.unleashLifeRemaining = false;
+        debug && console.log("Cast:", spellId);
       }
     }
   }
 
-  on_byPlayer_applybuff(event) {
+  on_byPlayer_removebuff(event) {
```

Comment:
Mostly unrelated question, is there no event for removing the unleash life buff?